Detectors/CPV/reconstruction/src/FullCluster.cxx

```diff
@@ -0,0 +1,210 @@
+        if (it == jt) {
+          continue;
+        }
+        if (Geometry::areNeighbours((*it).absId, (*jt).absId) == 1) {
+          hasNeighbours = true;
+          break;
+        }
+      }
+      if (!hasNeighbours) { //Isolated digits are rare
+        it = mElementList.erase(it);
+        --mMulDigit;
+      } else {
+        mEnergy += (*it).energy;
+        ++it;
+      }
+    }
+  } else {
+    mEnergy = mElementList[0].energy;
+  }
+}
+//____________________________________________________________________________
+void FullCluster::evalLocalPosition()
+{
+  // Calculates the center of gravity in the local CPV-module coordinates
+  // Note that correction for non-perpendicular incidence will be applied later
+  // when vertex will be known.
+
+  if (mEnergy <= 0.) { // zero energy cluster, position undefined
+    mLocalPosX = -999.;
+    mLocalPosZ = -999.;
+    return;
+  }
+
+  // find module number
+  mModule = Geometry::absIdToModule(mElementList[0].absId);
+
+  float wtot = 0.;
+  mLocalPosX = 0.;
+  mLocalPosZ = 0.;
+  float invE = 1. / mEnergy;
+  for (auto it : mElementList) {
+    float w = std::max(float(0.), o2::cpv::CPVSimParams::Instance().mLogWeight + std::log(it.energy * invE));
+    mLocalPosX += it.localX * w;
+    mLocalPosZ += it.localZ * w;
+    wtot += w;
+  }
+  if (wtot > 0) {
+    wtot = 1. / wtot;
+    mLocalPosX *= wtot;
+    mLocalPosZ *= wtot;
+  }
+}
+
+//____________________________________________________________________________
+std::vector<Digit>::const_iterator FullCluster::BinarySearch(const std::vector<Digit>* container, Digit& element)
+{
+  const std::vector<Digit>::const_iterator endIt = container->end();
+
+  std::vector<Digit>::const_iterator left = container->begin();
+  std::vector<Digit>::const_iterator right = endIt;
+
+  if (container->size() == 0 || container->front() > element || container->back() < element) {
+    return endIt;
+  }
+
+  while (distance(left, right) > 0) {
+    const std::vector<Digit>::const_iterator mid = left + distance(left, right) / 2;
+
+    if (element > *mid) {
+      left = mid + 1;
+    } else {
+      right = mid;
+    }
+  }
+
+  if (*right == element) {
+    return right;
+  }
+
+  return endIt;
+}
+//____________________________________________________________________________
+char FullCluster::getNumberOfLocalMax(gsl::span<int> maxAt) const
+{
+  // Calculates the number of local maxima in the cluster using LocalMaxCut as the minimum
+  // energy difference between maximum and surrounding digits
+
+  float locMaxCut = o2::cpv::CPVSimParams::Instance().mLocalMaximumCut;
+
+  std::unique_ptr<bool[]> isLocalMax = std::make_unique<bool[]>(mMulDigit);
+
+  for (int i = 0; i < mMulDigit; i++) {
+    if (mElementList[i].energy > o2::cpv::CPVSimParams::Instance().mClusteringThreshold) {
+      isLocalMax[i] = true;
+    } else {
+      isLocalMax[i] = false;
+    }
+  }
+  for (int i = 0; i < mMulDigit; i++) {
```

Comment:
this kind of loops is faster to do as
for (int i=mMulDigit;i--;) {
  for (int j=i;j--;) {
    .... // if the body does not depend on the order
 }
}